From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Mon, 23 May 2022 18:53:17 +0500
Subject: [PATCH] misc: rebrand to Reaper


diff --git a/pom.xml b/pom.xml
index 2fa2dd7c71e4f24f60e415c38ccf734c78713cbb..f5657f6f862387d593f74276ca512584b029b085 100644
--- a/pom.xml
+++ b/pom.xml
@@ -4,18 +4,18 @@
     <modelVersion>4.0.0</modelVersion>
 
     <parent>
-        <groupId>com.destroystokyo.paper</groupId>
-        <artifactId>paper-parent</artifactId>
+        <groupId>com.github.ruviolence</groupId>
+        <artifactId>reaper-parent</artifactId>
         <version>dev-SNAPSHOT</version>
     </parent>
 
-    <groupId>com.destroystokyo.paper</groupId>
-    <artifactId>paper-api</artifactId>
+    <groupId>com.github.ruviolence</groupId>
+    <artifactId>reaper-api</artifactId>
     <version>1.12.2-R0.1-SNAPSHOT</version>
     <packaging>jar</packaging>
 
-    <name>Paper-API</name>
-    <url>https://github.com/PaperMC/Paper</url>
+    <name>Reaper-API</name>
+    <url>https://github.com/ruViolence/Reaper</url>
     <description>An enhanced plugin API for Minecraft servers.</description>
 
     <properties>
diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index 6bb7607dde6c3f20f2f2b9b8b5dad877f84f6646..4fb24db889f77909380b7a8061adc03e573087de 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -209,8 +209,8 @@ public class VersionCommand extends BukkitCommand {
     private void obtainVersion() {
         String version = Bukkit.getVersion();
         if (version == null) version = "Custom";
-        if (version.startsWith("git-Paper-")) {
-            String[] parts = version.substring("git-Paper-".length()).split("[-\\s]");
+        if (version.startsWith("git-Reaper-")) {
+            String[] parts = version.substring("git-Reaper-".length()).split("[-\\s]");
             int distance = getDistance(null, parts[0]);
             switch (distance) {
                 case -1:
@@ -261,13 +261,8 @@ public class VersionCommand extends BukkitCommand {
 
     // Paper start
     private static int getDistance(String repo, String verInfo) {
-        try {
-            int currentVer = Integer.decode(verInfo);
-            return getFromJenkins(currentVer);
-        } catch (NumberFormatException ex) {
-            verInfo = verInfo.replace("\"", "");
-            return getFromRepo("PaperMC/Paper", verInfo);
-        }
+        verInfo = verInfo.replace("\"", "");
+        return getFromRepo("ruViolence/Reaper", verInfo);
             /*
             BufferedReader reader = Resources.asCharSource(
                     new URL("https://hub.spigotmc.org/stash/rest/api/1.0/projects/SPIGOT/repos/" + repo + "/commits?since=" + URLEncoder.encode(hash, "UTF-8") + "&withCounts=true"),
@@ -285,27 +280,6 @@ public class VersionCommand extends BukkitCommand {
             */
     }
 
-    private static int getFromJenkins(int currentVer) {
-        try {
-            BufferedReader reader = Resources.asCharSource(
-                    new URL("https://ci.destroystokyo.com/job/Paper/lastSuccessfulBuild/buildNumber"), // Paper
-                    Charsets.UTF_8
-            ).openBufferedStream();
-            try {
-                int newVer = Integer.decode(reader.readLine());
-                return newVer - currentVer;
-            } catch (NumberFormatException ex) {
-                ex.printStackTrace();
-                return -2;
-            } finally {
-                reader.close();
-            }
-        } catch (IOException e) {
-            e.printStackTrace();
-            return -1;
-        }
-    }
-
     // Contributed by Techcable <Techcable@outlook.com> in GH PR #65
     private static final String BRANCH = "ver/1.12.2";
     private static int getFromRepo(String repo, String hash) {
