From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Sat, 4 Jun 2022 00:32:33 +0500
Subject: [PATCH] perf: don't tick end gateway if not in the end world


diff --git a/src/main/java/net/minecraft/server/TileEntityEndGateway.java b/src/main/java/net/minecraft/server/TileEntityEndGateway.java
index eb2dbbe53498177d2ba45c32068ee7baacb7d51a..df9ce67f4010a9359b62f4dac3615cc1f9c44023 100644
--- a/src/main/java/net/minecraft/server/TileEntityEndGateway.java
+++ b/src/main/java/net/minecraft/server/TileEntityEndGateway.java
@@ -48,6 +48,7 @@ public class TileEntityEndGateway extends TileEntityEnderPortal implements ITick
     }
 
     public void e() {
+        if (!(this.world.worldProvider instanceof WorldProviderTheEnd)) return; // Reaper - Don't tick if not in the end world
         boolean flag = this.a();
         boolean flag1 = this.f();
 
