From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Fri, 3 Jun 2022 03:21:58 +0500
Subject: [PATCH] perf: cache world configs


diff --git a/src/main/java/com/destroystokyo/paper/PaperCommand.java b/src/main/java/com/destroystokyo/paper/PaperCommand.java
index 5615a019962dc08ee2aa9833e46504d14e1abeab..41d6fbc96c874fc84d4d9a08e99180468d487d5a 100644
--- a/src/main/java/com/destroystokyo/paper/PaperCommand.java
+++ b/src/main/java/com/destroystokyo/paper/PaperCommand.java
@@ -187,6 +187,7 @@ public class PaperCommand extends Command {
         for (WorldServer world : console.worlds) {
             world.paperConfig.init();
         }
+        com.github.ruviolence.reaper.WorldConfigCache.dropCache(); // Reaper - Cache world configs
         console.server.reloadCount++;
 
         Command.broadcastCommandMessage(sender, ChatColor.GREEN + "Paper config reload complete.");
diff --git a/src/main/java/com/github/ruviolence/reaper/WorldConfigCache.java b/src/main/java/com/github/ruviolence/reaper/WorldConfigCache.java
new file mode 100644
index 0000000000000000000000000000000000000000..e556587cf45f9fe460c10b57a6071cf2418a828f
--- /dev/null
+++ b/src/main/java/com/github/ruviolence/reaper/WorldConfigCache.java
@@ -0,0 +1,26 @@
+package com.github.ruviolence.reaper;
+
+import com.destroystokyo.paper.PaperWorldConfig;
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import org.spigotmc.SpigotWorldConfig;
+
+import java.util.concurrent.TimeUnit;
+
+public class WorldConfigCache {
+    private static final Cache<String, SpigotWorldConfig> spigotWorldConfigCache = Caffeine.newBuilder().expireAfterAccess(30, TimeUnit.MINUTES).build();
+    private static final Cache<String, PaperWorldConfig> paperWorldConfigCache = Caffeine.newBuilder().expireAfterAccess(30, TimeUnit.MINUTES).build();
+
+    public static SpigotWorldConfig getSpigotConfig(String worldName) {
+        return spigotWorldConfigCache.get(worldName, SpigotWorldConfig::new);
+    }
+
+    public static PaperWorldConfig getPaperConfig(String worldName, SpigotWorldConfig spigotWorldConfig) {
+        return paperWorldConfigCache.get(worldName, name -> new PaperWorldConfig(name, spigotWorldConfig));
+    }
+
+    public static void dropCache() {
+        spigotWorldConfigCache.invalidateAll();
+        paperWorldConfigCache.invalidateAll();
+    }
+}
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 354579f42865d4d989cce1a6889632581068067d..a5fe16980ba63e645f240632b16beae282c26e5c 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -174,8 +174,8 @@ public abstract class World implements IBlockAccess {
     }
 
     protected World(IDataManager idatamanager, WorldData worlddata, WorldProvider worldprovider, MethodProfiler methodprofiler, boolean flag, ChunkGenerator gen, org.bukkit.World.Environment env) {
-        this.spigotConfig = new org.spigotmc.SpigotWorldConfig( worlddata.getName() ); // Spigot
-        this.paperConfig = new com.destroystokyo.paper.PaperWorldConfig(worlddata.getName(), this.spigotConfig); // Paper
+        this.spigotConfig = com.github.ruviolence.reaper.WorldConfigCache.getSpigotConfig( worlddata.getName() ); // Spigot // Reaper - Cache world configs
+        this.paperConfig = com.github.ruviolence.reaper.WorldConfigCache.getPaperConfig(worlddata.getName(), this.spigotConfig); // Paper // Reaper - Cache world configs
         this.generator = gen;
         this.world = new CraftWorld((WorldServer) this, gen, env);
         this.ticksPerAnimalSpawns = this.getServer().getTicksPerAnimalSpawns(); // CraftBukkit
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 95bc3e2e328c9faf79ef9a65a47c7b779708d564..b800048737aa316548cd968a22ed81d6a566089e 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -760,6 +760,7 @@ public final class CraftServer implements Server {
             }
             world.spigotConfig.init(); // Spigot
             world.paperConfig.init(); // Paper
+            com.github.ruviolence.reaper.WorldConfigCache.dropCache(); // Reaper - Cache world configs
         }
 
         Plugin[] pluginClone = pluginManager.getPlugins().clone(); // Paper
diff --git a/src/main/java/org/spigotmc/SpigotCommand.java b/src/main/java/org/spigotmc/SpigotCommand.java
index fdb030c7af3a63769302391a0aab6ad10f77fb30..f0730d03fca2a21157eb75b8e6904fac3c963892 100644
--- a/src/main/java/org/spigotmc/SpigotCommand.java
+++ b/src/main/java/org/spigotmc/SpigotCommand.java
@@ -34,6 +34,7 @@ public class SpigotCommand extends Command {
             for (WorldServer world : console.worlds) {
                 world.spigotConfig.init();
             }
+            com.github.ruviolence.reaper.WorldConfigCache.dropCache(); // Reaper - Cache world configs
             console.server.reloadCount++;
 
             Command.broadcastCommandMessage(sender, ChatColor.GREEN + "Reload complete.");
