From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Wed, 8 Jun 2022 11:04:53 +0500
Subject: [PATCH] fix: PlayerInteractEvent not called in specific situations


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 8b4ee13af03964c54ba85cacd3e46a7db65efd90..0056a4b2011ea73c5c91a8d310b600de70a90ab1 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1052,8 +1052,10 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                 org.bukkit.event.player.PlayerInteractEvent event = CraftEventFactory.callPlayerInteractEvent(this.player, Action.RIGHT_CLICK_AIR, itemstack, enumhand);
                 cancelled = event.useItemInHand() == Event.Result.DENY;
             } else {
-                if (player.playerInteractManager.firedInteract) {
-                    player.playerInteractManager.firedInteract = false;
+                // Reaper start
+                if (player.playerInteractManager.firedInteractTick == MinecraftServer.currentTick && itemstack == player.playerInteractManager.firedInteractItem) {
+                    player.playerInteractManager.firedInteractTick = 0;
+                    // Reaper end
                     cancelled = player.playerInteractManager.interactResult;
                 } else {
                     org.bukkit.event.player.PlayerInteractEvent event = CraftEventFactory.callPlayerInteractEvent(player, Action.RIGHT_CLICK_BLOCK, movingobjectposition.a(), movingobjectposition.direction, itemstack, true, enumhand);
diff --git a/src/main/java/net/minecraft/server/PlayerInteractManager.java b/src/main/java/net/minecraft/server/PlayerInteractManager.java
index 0bc3aa6fc30de57fa87d51f17ae20360ba13b037..61c4a1e06bc3e8deee52ec1994d078c1aa5c1488 100644
--- a/src/main/java/net/minecraft/server/PlayerInteractManager.java
+++ b/src/main/java/net/minecraft/server/PlayerInteractManager.java
@@ -463,7 +463,10 @@ public class PlayerInteractManager {
 
     // CraftBukkit start - whole method
     public boolean interactResult = false;
-    public boolean firedInteract = false;
+    // Reaper start
+    public int firedInteractTick = 0;
+    public ItemStack firedInteractItem = null;
+    // Reaper end
     public EnumInteractionResult a(EntityHuman entityhuman, World world, ItemStack itemstack, EnumHand enumhand, BlockPosition blockposition, EnumDirection enumdirection, float f, float f1, float f2) {
         IBlockData blockdata = world.getType(blockposition);
         EnumInteractionResult enuminteractionresult = EnumInteractionResult.FAIL;
@@ -488,7 +491,10 @@ public class PlayerInteractManager {
             }
 
             PlayerInteractEvent event = CraftEventFactory.callPlayerInteractEvent(entityhuman, Action.RIGHT_CLICK_BLOCK, blockposition, enumdirection, itemstack, cancelledBlock, enumhand);
-            firedInteract = true;
+            // Reaper start
+            firedInteractTick = MinecraftServer.currentTick;
+            firedInteractItem = itemstack;
+            // Reaper end
             interactResult = event.useItemInHand() == Event.Result.DENY;
 
             if (event.useInteractedBlock() == Event.Result.DENY) {
