From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Wed, 8 Jun 2022 13:43:28 +0500
Subject: [PATCH] fix: the world not saving


diff --git a/src/main/java/net/minecraft/server/ChunkProviderServer.java b/src/main/java/net/minecraft/server/ChunkProviderServer.java
index 5833e5d8e340efc706bc10472fd61b6f224f87c3..5ada3fb0ccec92e8c6d090e2e3dd8f27b4695db0 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderServer.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderServer.java
@@ -299,6 +299,7 @@ public class ChunkProviderServer implements IChunkProvider {
             }
         }
 
+        if (!flag) // Reaper - Fix the world not saving
         if (queueSize > world.paperConfig.queueSizeAutoSaveThreshold){
             return false;
         }
