From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Sat, 28 May 2022 14:21:29 +0500
Subject: [PATCH] fix: entityliving double dying

Fixes some dupes

diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 3daf3d05fdd319bf7c798ec8e826c3eeb371affa..10857ed16dc5d1d526ae7248edcd3f76393836c9 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -1125,6 +1125,7 @@ public abstract class EntityLiving extends Entity {
     }
 
     public void die(DamageSource damagesource) {
+        if (this.dead) return; // Reaper
         if (!this.aU) {
             Entity entity = damagesource.getEntity();
             EntityLiving entityliving = this.ci();
