From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Sat, 28 May 2022 14:21:29 +0500
Subject: [PATCH] fix: entity portal dupe


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 8061a7c12aa32e6dc9bdd697983925cfd93eee96..7d647b202d022bc575ad86394993702f30664741 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -2562,7 +2562,7 @@ public abstract class Entity implements ICommandListener, KeyedObject { // Paper
             // CraftBukkit end */
             // CraftBukkit start - Ensure chunks are loaded in case TravelAgent is not used which would initially cause chunks to load during find/create
             // minecraftserver.getPlayerList().changeWorld(this, j, worldserver, worldserver1);
-            worldserver1.getMinecraftServer().getPlayerList().repositionEntity(this, exit, portal);
+            // worldserver1.getMinecraftServer().getPlayerList().repositionEntity(this, exit, portal); // Reaper - Moved down
             // worldserver.entityJoinedWorld(this, false); // Handled in repositionEntity
             // CraftBukkit end
             // this.world.methodProfiler.c("reloading"); // Reaper - Remove MethodProfiler
@@ -2570,6 +2570,7 @@ public abstract class Entity implements ICommandListener, KeyedObject { // Paper
 
             if (entity != null) {
                 entity.a(this);
+                worldserver1.getMinecraftServer().getPlayerList().repositionEntity(this, exit, portal); // Reaper
                 /* CraftBukkit start - We need to do this...
                 if (j == 1 && i == 1) {
                     BlockPosition blockposition1 = worldserver1.q(worldserver1.getSpawn());
diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 108c16c5915ea70290056c4581f16869910093e9..ad2aa3ab4249ddca9c7812c8a395fe50653da3e4 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -1125,6 +1125,7 @@ public abstract class EntityLiving extends Entity {
     }
 
     public void die(DamageSource damagesource) {
+        if (this.dead) return; // Reaper
         if (!this.aU) {
             Entity entity = damagesource.getEntity();
             EntityLiving entityliving = this.ci();
