From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Sun, 19 Jun 2022 22:13:40 +0500
Subject: [PATCH] fix: do not reattach RootVehicle that too far


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 1ffd14cdd241c590eb89fb4a35959a75424e0dde..eb3f6df703416c0a9c26b68eb841c3ae8f835d99 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -209,6 +209,7 @@ public abstract class PlayerList {
                 Iterator iterator1;
                 Entity entity1;
 
+                if (entityplayer.h(entity) < 256.0D) // Reaper
                 if (entity.getUniqueID().equals(uuid)) {
                     entityplayer.a(entity, true);
                 } else {
