From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Sun, 19 Jun 2022 22:13:40 +0500
Subject: [PATCH] fix: do not reattach RootVehicle that too far


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index bf6c511f3ce8305b22f2e460e2dd1cf4106edf7a..9cb528a391068ee2fe16c216fa66a1cfb0361768 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -209,6 +209,7 @@ public abstract class PlayerList {
                 Iterator iterator1;
                 Entity entity1;
 
+                if (entityplayer.h(entity) < 256.0D) // Reaper
                 if (entity.getUniqueID().equals(uuid)) {
                     entityplayer.a(entity, true);
                 } else {
