From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Fri, 27 May 2022 01:31:12 +0500
Subject: [PATCH] feat: always full chat visibility


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 43ff36f6e2205cb2bcd785abe663c151a2572c22..c53cd688c58b6355c3837b461945fca80792e12f 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -1320,7 +1320,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
             this.server.server.getPluginManager().callEvent(event);
         }
         // CraftBukkit end
-        this.cl = packetplayinsettings.c();
+        this.cl = EnumChatVisibility.FULL; // Reaper - Always full chat visibility
         this.cm = packetplayinsettings.d();
         this.getDataWatcher().set(EntityPlayer.br, Byte.valueOf((byte) packetplayinsettings.e()));
         this.getDataWatcher().set(EntityPlayer.bs, Byte.valueOf((byte) (packetplayinsettings.getMainHand() == EnumMainHand.LEFT ? 0 : 1)));
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 78331c8e9abe7a615fe7484d2cf371d9b96eddd1..06f8da246caf5e85cafd0dbb3b2ecb80483c09c6 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1195,6 +1195,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
     }
 
     public void sendPacket(final Packet<?> packet) {
+        /* // Reaper start - Always full chat visibility
         if (packet instanceof PacketPlayOutChat) {
             PacketPlayOutChat packetplayoutchat = (PacketPlayOutChat) packet;
             EntityHuman.EnumChatVisibility entityhuman_enumchatvisibility = this.player.getChatFlags();
@@ -1207,6 +1208,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                 return;
             }
         }
+        */ // Reaper end
 
         // CraftBukkit start
         if (packet == null || this.processedDisconnect) { // Spigot
@@ -1268,7 +1270,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
             PlayerConnectionUtils.ensureMainThread(packetplayinchat, this, this.player.x());
         }
         // CraftBukkit end
-        if (this.player.dead || this.player.getChatFlags() == EntityHuman.EnumChatVisibility.HIDDEN) { // CraftBukkit - dead men tell no tales
+        if (this.player.dead /*|| this.player.getChatFlags() == EntityHuman.EnumChatVisibility.HIDDEN*/) { // CraftBukkit - dead men tell no tales // Reaper - Always full chat visibility
             ChatMessage chatmessage = new ChatMessage("chat.cannotSend", new Object[0]);
 
             chatmessage.getChatModifier().setColor(EnumChatFormat.RED);
@@ -1331,11 +1333,13 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                     }
                 } );
                 // Spigot end
+                /* // Reaper start - Always full chat visibility
             } else if (this.player.getChatFlags() == EntityHuman.EnumChatVisibility.SYSTEM) { // Re-add "Command Only" flag check
                 ChatMessage chatmessage = new ChatMessage("chat.cannotSend", new Object[0]);
 
                 chatmessage.getChatModifier().setColor(EnumChatFormat.RED);
                 this.sendPacket(new PacketPlayOutChat(chatmessage));
+                */ // Reaper end
             } else if (true) {
                 this.chat(s, true);
                 // CraftBukkit end - the below is for reference. :)
@@ -1388,7 +1392,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
 
     // CraftBukkit start - add method
     public void chat(String s, boolean async) {
-        if (s.isEmpty() || this.player.getChatFlags() == EntityHuman.EnumChatVisibility.HIDDEN) {
+        if (s.isEmpty() /*|| this.player.getChatFlags() == EntityHuman.EnumChatVisibility.HIDDEN*/) { // Reaper - Always full chat visibility
             return;
         }
 
@@ -1417,7 +1421,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
             }
             // Paper End
             this.handleCommand(s);
-        } else if (this.player.getChatFlags() == EntityHuman.EnumChatVisibility.SYSTEM) {
+        } else if (/*this.player.getChatFlags() == EntityHuman.EnumChatVisibility.SYSTEM*/false) { // Reaper - Always full chat visibility
             // Do nothing, this is coming from a plugin
         } else {
             Player player = this.getPlayer();
