From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Thu, 2 Jun 2022 23:27:42 +0500
Subject: [PATCH] feat: add AnyEntitySpawnEvent


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index bbc1a9ec5055480468caf25f20fc430c2671a023..7bccdebea8c971c790687f0ece6f4b94162ce0ff 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1223,6 +1223,10 @@ public abstract class World implements IBlockAccess {
                 } // Paper end - End iteration skip check - All tweaking ends here
             }
         } // Spigot end
+        // Reaper start
+        else if (com.github.ruviolence.reaper.event.entity.AnyEntitySpawnEvent.getHandlerList().getRegisteredListeners().length != 0) {
+            event = CraftEventFactory.callAnyEntitySpawnEvent(entity, spawnReason);
+        } // Reaper end
 
         if (event != null && (event.isCancelled() || entity.dead)) {
             entity.dead = true;
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 9727853a41754a14416611b5184bcd7a3cb02d28..f9e25970d40db707576c7d8a76a27b39d7985d5d 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -75,6 +75,7 @@ public class CraftEventFactory {
     public static Entity entityDamage; // For use in EntityDamageByEntityEvent
     // Reaper start
     private static final com.github.ruviolence.reaper.event.entity.EntityMoveEvent entityMoveEvent = new com.github.ruviolence.reaper.event.entity.EntityMoveEvent(null);
+    private static final com.github.ruviolence.reaper.event.entity.AnyEntitySpawnEvent anyEntitySpawnEvent = new com.github.ruviolence.reaper.event.entity.AnyEntitySpawnEvent(null);
     // Reaper end
 
     // helper methods
@@ -1192,5 +1193,17 @@ public class CraftEventFactory {
 
         return !event.isCancelled();
     }
+
+    public static com.github.ruviolence.reaper.event.entity.AnyEntitySpawnEvent callAnyEntitySpawnEvent(Entity entity, SpawnReason spawnReason) {
+        com.github.ruviolence.reaper.event.entity.AnyEntitySpawnEvent event = CraftEventFactory.anyEntitySpawnEvent;
+
+        event.setData(
+                entity.getBukkitEntity(),
+                spawnReason);
+
+        Bukkit.getPluginManager().callEvent(event);
+
+        return event;
+    }
     // Reaper end
 }
