From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Fri, 27 May 2022 08:16:09 +0500
Subject: [PATCH] feat: configurable portal travel cache


diff --git a/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java b/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java
index 726b89f392de6351484e56a3799ffc588ab176a4..9b7a1ce6caf7ac513d85174817fb0173a8b2ad42 100644
--- a/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java
+++ b/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java
@@ -123,4 +123,11 @@ public class ReaperConfig {
         config.addDefault(path, def);
         return config.getString(path, config.getString(path));
     }
+
+    public static int portalTravelCacheLife;
+    public static boolean portalTravelInstantExpire;
+    private static void portalTravel() {
+        portalTravelCacheLife = 20 * getInt("portal-travel.cache-life", 15);
+        portalTravelInstantExpire = getBoolean("portal-travel.instant-expire", false);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/BlockPortal.java b/src/main/java/net/minecraft/server/BlockPortal.java
index 5deae342ad17166b41d88e58f4da7d30d7250449..efd57833662ec1f3cb69b6ac82c66e3ea16449ec 100644
--- a/src/main/java/net/minecraft/server/BlockPortal.java
+++ b/src/main/java/net/minecraft/server/BlockPortal.java
@@ -98,11 +98,13 @@ public class BlockPortal extends BlockHalfTransparent {
             blockportal_shape = new BlockPortal.Shape(world, blockposition, EnumDirection.EnumAxis.X);
             if (!blockportal_shape.d() || blockportal_shape.e < blockportal_shape.width * blockportal_shape.height) {
                 world.setTypeUpdate(blockposition, Blocks.AIR.getBlockData());
+                if (com.github.ruviolence.reaper.ReaperConfig.portalTravelInstantExpire) ((WorldServer) world).getTravelAgent().expirePortal(blockposition); // Reaper - Configurable portal travel cache
             }
         } else if (enumdirection_enumaxis == EnumDirection.EnumAxis.Z) {
             blockportal_shape = new BlockPortal.Shape(world, blockposition, EnumDirection.EnumAxis.Z);
             if (!blockportal_shape.d() || blockportal_shape.e < blockportal_shape.width * blockportal_shape.height) {
                 world.setTypeUpdate(blockposition, Blocks.AIR.getBlockData());
+                if (com.github.ruviolence.reaper.ReaperConfig.portalTravelInstantExpire) ((WorldServer) world).getTravelAgent().expirePortal(blockposition); // Reaper - Configurable portal travel cache
             }
         }
 
diff --git a/src/main/java/net/minecraft/server/PortalTravelAgent.java b/src/main/java/net/minecraft/server/PortalTravelAgent.java
index 3c6f6f11312632a29a20c8278d1680d63babb0d8..897a92d1c193c3b33a679fbd4cded878710e1ed7 100644
--- a/src/main/java/net/minecraft/server/PortalTravelAgent.java
+++ b/src/main/java/net/minecraft/server/PortalTravelAgent.java
@@ -493,7 +493,7 @@ public class PortalTravelAgent {
 
     public void a(long i) {
         if (i % 100L == 0L) {
-            long j = i - 300L;
+            long j = i - com.github.ruviolence.reaper.ReaperConfig.portalTravelCacheLife; // Reaper - Configurable portal travel cache
             ObjectIterator objectiterator = this.c.values().iterator();
 
             while (objectiterator.hasNext()) {
@@ -506,6 +506,16 @@ public class PortalTravelAgent {
         }
 
     }
+    
+    // Reaper start - Configurable portal travel cache
+    public void expirePortal(BlockPosition portal) {
+        for (ObjectIterator<ChunkCoordinatesPortal> iterator = this.c.values().iterator(); iterator.hasNext(); ) {
+            if (iterator.next().equals(portal)) {
+                iterator.remove();
+            }
+        }
+    }
+    // Reaper end
 
     public class ChunkCoordinatesPortal extends BlockPosition {
 
