From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Thu, 2 Jun 2022 08:04:40 +0500
Subject: [PATCH] feat: configurable step sounds


diff --git a/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java b/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java
index d906587c1a68a65f5308a948f009a05ec270b299..0738b4d7907d850c520cc89f35ba0e74515ca5de 100644
--- a/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java
+++ b/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java
@@ -181,4 +181,9 @@ public class ReaperConfig {
     private static void paperCustomFlowSpeed() {
         paperCustomFlowSpeed = getBoolean("paper-custom-flow-speed", true);
     }
+
+    public static boolean stepSounds;
+    private static void stepSounds() {
+        stepSounds = getBoolean("step-sounds", true);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 0e23dc56d2e9e45785070a23063ff6463fa330e9..180d5cde8dd751cca950040dc9a9c8d8de8922fb 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -954,6 +954,7 @@ public abstract class Entity implements ICommandListener, KeyedObject { // Paper
 
                         this.a(this.ae(), f1, 1.0F + (this.random.nextFloat() - this.random.nextFloat()) * 0.4F);
                     } else {
+                        if (com.github.ruviolence.reaper.ReaperConfig.stepSounds) // Reaper
                         this.a(blockposition, block1);
                     }
                 } else if (this.K > this.ay && this.ah() && iblockdata.getMaterial() == Material.AIR) {
