From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Fri, 10 Jun 2022 16:06:05 +0500
Subject: [PATCH] feat: round player ping

Helps to hide the real ping from cheat clients.

diff --git a/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java b/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java
index a7ed9e0614697ade04b65761a04c94dff6ba63ac..5a2ac5cfa691f7a63b7c359fe4cbad413822e335 100644
--- a/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java
+++ b/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java
@@ -259,4 +259,9 @@ public class ReaperConfig {
     private static void relativeArrowMotion() {
         relativeArrowMotion = getBoolean("relative-arrow-motion", true);
     }
+
+    public static boolean roundPlayerPing;
+    private static void roundPlayerPing() {
+        roundPlayerPing = getBoolean("round-player-ping", false);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java b/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java
index d385a9c950fe3b955536b88cfc6ce8be27cdc587..f244441f879368a0d71751e6163e1d864dabb8a6 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java
@@ -192,6 +192,22 @@ public class PacketPlayOutPlayerInfo implements Packet<PacketListenerPlayOut> {
         }
 
         public int b() {
+            // Reaper start
+            if (com.github.ruviolence.reaper.ReaperConfig.roundPlayerPing) {
+                int ping = this.b;
+
+                if (ping < 150) {
+                    return 0;
+                } else if (ping < 300) {
+                    return 150;
+                } else if (ping < 600) {
+                    return 300;
+                } else if (ping < 1000) {
+                    return 600;
+                } else {
+                    return 1000;
+                }
+            } // Reaper end
             return this.b;
         }
 
