From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Thu, 14 Jul 2022 05:11:53 +0500
Subject: [PATCH] perf: don't load chunks for physics


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 29eb8c5ed591670021c9d246a4516796a5adc2e9..7f63290cd7b6c649da9114a957bb6aa7493f966d 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -677,7 +677,10 @@ public abstract class World implements IBlockAccess {
 
     public void b(BlockPosition blockposition, final Block block, BlockPosition blockposition1) {
         if (true) { // Reaper - Remove isClientSide check
-            IBlockData iblockdata = this.getType(blockposition);
+            // Reaper start - Don't load chunks for physics
+            IBlockData iblockdata = this.getTypeIfLoaded(blockposition);
+            if (iblockdata == null) return;
+            // Reaper end
 
             if (iblockdata.getBlock() == Blocks.dk) {
                 try {
