From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Thu, 14 Jul 2022 05:13:06 +0500
Subject: [PATCH] perf: don't load chunks for light checks


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index b5a4f020de60e9826c146f1e590aaeb792dbc247..3a20967dea557b617b3801a7ac8254d40e6cd835 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -747,7 +747,13 @@ public abstract class World implements IBlockAccess {
                 blockposition = new BlockPosition(blockposition.getX(), 255, blockposition.getZ());
             }
 
-            return this.getChunkAtWorldCoords(blockposition).a(blockposition, 0);
+            // Reaper start - Don't load chunks for light checks
+            Chunk chunk = this.getChunkIfLoaded(blockposition);
+            if (chunk == null) {
+                return 0;
+            }
+            return chunk.a(blockposition, 0);
+            // Reaper end
         }
     }
 
