From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Thu, 14 Jul 2022 05:17:27 +0500
Subject: [PATCH] perf: don't load chunks for endermans


diff --git a/src/main/java/net/minecraft/server/EntityEnderman.java b/src/main/java/net/minecraft/server/EntityEnderman.java
index 26f08a7be0a49313ce0596a26f65410e0de3ca0d..65f88cd5209f8581842cefd076daffbb47a6bc7a 100644
--- a/src/main/java/net/minecraft/server/EntityEnderman.java
+++ b/src/main/java/net/minecraft/server/EntityEnderman.java
@@ -342,7 +342,10 @@ public class EntityEnderman extends EntityMonster {
             int j = MathHelper.floor(this.enderman.locY + random.nextDouble() * 3.0D);
             int k = MathHelper.floor(this.enderman.locZ - 2.0D + random.nextDouble() * 4.0D);
             BlockPosition blockposition = new BlockPosition(i, j, k);
-            IBlockData iblockdata = world.getType(blockposition);
+            // Reaper start - Don't load chunks for endermans
+            IBlockData iblockdata = world.getTypeIfLoaded(blockposition);
+            if (iblockdata == null) return;
+            // Reaper end
             Block block = iblockdata.getBlock();
             MovingObjectPosition movingobjectposition = world.rayTrace(new Vec3D((double) ((float) MathHelper.floor(this.enderman.locX) + 0.5F), (double) ((float) j + 0.5F), (double) ((float) MathHelper.floor(this.enderman.locZ) + 0.5F)), new Vec3D((double) ((float) i + 0.5F), (double) ((float) j + 0.5F), (double) ((float) k + 0.5F)), false, true, false);
             boolean flag = movingobjectposition != null && movingobjectposition.a().equals(blockposition);
@@ -378,7 +381,10 @@ public class EntityEnderman extends EntityMonster {
             int j = MathHelper.floor(this.a.locY + random.nextDouble() * 2.0D);
             int k = MathHelper.floor(this.a.locZ - 1.0D + random.nextDouble() * 2.0D);
             BlockPosition blockposition = new BlockPosition(i, j, k);
-            IBlockData iblockdata = world.getType(blockposition);
+            // Reaper start - Don't load chunks for endermans
+            IBlockData iblockdata = world.getTypeIfLoaded(blockposition);
+            if (iblockdata == null) return;
+            // Reaper end
             IBlockData iblockdata1 = world.getType(blockposition.down());
             IBlockData iblockdata2 = this.a.getCarried();
 
