From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Thu, 14 Jul 2022 06:21:12 +0500
Subject: [PATCH] perf: cache block break animation packet


diff --git a/src/main/java/net/minecraft/server/WorldManager.java b/src/main/java/net/minecraft/server/WorldManager.java
index 49dfc7ea0fe28ad8a8a23050023adb97ea56eb94..6f1a8bc5db6ce8f3ba28060748a3e80476b605cd 100644
--- a/src/main/java/net/minecraft/server/WorldManager.java
+++ b/src/main/java/net/minecraft/server/WorldManager.java
@@ -70,6 +70,7 @@ public class WorldManager implements IWorldAccess {
         // Paper start
         java.util.List<? extends EntityHuman> list = entity != null ? entity.world.players : this.a.getPlayerList().v();
         Iterator<? extends EntityHuman> iterator = list.iterator();
+        PacketPlayOutBlockBreakAnimation packet = null; // Reaper - Cache block break animation packet
         while (iterator.hasNext()) {
             EntityHuman human = iterator.next();
             if (!(human instanceof EntityPlayer)) continue;
@@ -88,7 +89,10 @@ public class WorldManager implements IWorldAccess {
                 // CraftBukkit end
 
                 if (d0 * d0 + d1 * d1 + d2 * d2 < 1024.0D) {
-                    entityplayer.playerConnection.sendPacket(new PacketPlayOutBlockBreakAnimation(i, blockposition, j));
+                    // Reaper start - Cache block break animation packet
+                    if (packet == null) packet = new PacketPlayOutBlockBreakAnimation(i, blockposition, j);
+                    entityplayer.playerConnection.sendPacket(packet);
+                    // Reaper end
                 }
             }
         }
