From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Thu, 14 Jul 2022 06:31:14 +0500
Subject: [PATCH] perf: skip updating entity tracker without players


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 960db455d58a90b9ec9895b4d81c1b36dfd6feb6..5fcba6969ea9a2fe899f86ee838a97f468883d9d 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -979,6 +979,7 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IAs
                     }
                 }
                 try { // Reaper end - Flush consolidation
+                if (this.getPlayerList().getPlayerCount() != 0) // Reaper - Skip updating entity tracker without players
                 worldserver.getTracker().updatePlayers();
                 // Reaper start - Flush consolidation
                 } finally {
