From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Thu, 14 Jul 2022 06:33:20 +0500
Subject: [PATCH] fix: reset players airTicks on respawn


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 8820b593a060fc47aa8bee3ba4e057b99ed2624a..19a04cec758504f7fb0f506804f5cf2552d08993 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -1639,6 +1639,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
 
         this.setHealth(this.getMaxHealth());
         this.fireTicks = 0;
+        this.setAirTicks(this.aM()); // Reaper - Reset airTicks on respawn
         this.fallDistance = 0;
         this.foodData = new FoodMetaData(this);
         this.expLevel = this.newLevel;
