From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Fri, 15 Jul 2022 04:36:43 +0500
Subject: [PATCH] fix-cow-dupe


diff --git a/src/main/java/net/minecraft/server/EntityAgeable.java b/src/main/java/net/minecraft/server/EntityAgeable.java
index f89bce9dba8a902d5c9a3d08617f5a4ce946744e..831ced608ceb509503573bd0d0b73a568448fd0c 100644
--- a/src/main/java/net/minecraft/server/EntityAgeable.java
+++ b/src/main/java/net/minecraft/server/EntityAgeable.java
@@ -44,7 +44,7 @@ public abstract class EntityAgeable extends EntityCreature {
     @Nullable
     public abstract EntityAgeable createChild(EntityAgeable entityageable);
 
-    public boolean a(EntityHuman entityhuman, EnumHand enumhand) {
+    public boolean a(EntityHuman entityhuman, EnumHand enumhand) throws InterruptedException {
         ItemStack itemstack = entityhuman.b(enumhand);
 
         if (itemstack.getItem() == Items.SPAWN_EGG) {
diff --git a/src/main/java/net/minecraft/server/EntityAnimal.java b/src/main/java/net/minecraft/server/EntityAnimal.java
index 54697148c6fefbaf94a243d3da72ebc3a2741199..22f3dc05ab85a4823026adf1ddb8036e9f97d4e3 100644
--- a/src/main/java/net/minecraft/server/EntityAnimal.java
+++ b/src/main/java/net/minecraft/server/EntityAnimal.java
@@ -105,7 +105,7 @@ public abstract class EntityAnimal extends EntityAgeable implements IAnimal {
         return itemstack.getItem() == Items.WHEAT;
     }
 
-    public boolean a(EntityHuman entityhuman, EnumHand enumhand) {
+    public boolean a(EntityHuman entityhuman, EnumHand enumhand) throws InterruptedException {
         ItemStack itemstack = entityhuman.b(enumhand);
 
         if (!itemstack.isEmpty()) {
diff --git a/src/main/java/net/minecraft/server/EntityCow.java b/src/main/java/net/minecraft/server/EntityCow.java
index 340b4ca3f6591d26ffb3919a57c0275ba58285aa..7b628b5efaf51ff6307ce28cc3742bb959e038ae 100644
--- a/src/main/java/net/minecraft/server/EntityCow.java
+++ b/src/main/java/net/minecraft/server/EntityCow.java
@@ -59,7 +59,7 @@ public class EntityCow extends EntityAnimal {
         return LootTables.L;
     }
 
-    public boolean a(EntityHuman entityhuman, EnumHand enumhand) {
+    public boolean a(EntityHuman entityhuman, EnumHand enumhand) throws InterruptedException {
         ItemStack itemstack = entityhuman.b(enumhand);
 
         if (itemstack.getItem() == Items.BUCKET && !entityhuman.abilities.canInstantlyBuild && !this.isBaby()) {
diff --git a/src/main/java/net/minecraft/server/EntityInsentient.java b/src/main/java/net/minecraft/server/EntityInsentient.java
index 28f5262785269be761272ba51f7e172d7016a780..0517bd9d10aec0fd4f7d6ffbc7354bdb45fd0982 100644
--- a/src/main/java/net/minecraft/server/EntityInsentient.java
+++ b/src/main/java/net/minecraft/server/EntityInsentient.java
@@ -1053,7 +1053,7 @@ public abstract class EntityInsentient extends EntityLiving {
         }
     }
 
-    protected boolean a(EntityHuman entityhuman, EnumHand enumhand) {
+    protected boolean a(EntityHuman entityhuman, EnumHand enumhand) throws InterruptedException {
         return false;
     }
 
diff --git a/src/main/java/net/minecraft/server/EntityMushroomCow.java b/src/main/java/net/minecraft/server/EntityMushroomCow.java
index 8bbbade37202405699e7048cdfae323dc152bb90..0a3ea3daddc39d34811e8898b6762627f21964ba 100644
--- a/src/main/java/net/minecraft/server/EntityMushroomCow.java
+++ b/src/main/java/net/minecraft/server/EntityMushroomCow.java
@@ -5,6 +5,7 @@ import org.bukkit.event.player.PlayerShearEntityEvent; // CraftBukkit
 
 public class EntityMushroomCow extends EntityCow {
 
+
     public EntityMushroomCow(World world) {
         super(world);
         this.setSize(0.9F, 1.4F);
@@ -15,7 +16,7 @@ public class EntityMushroomCow extends EntityCow {
         EntityInsentient.a(dataconvertermanager, EntityMushroomCow.class);
     }
 
-    public boolean a(EntityHuman entityhuman, EnumHand enumhand) {
+    public boolean a(EntityHuman entityhuman, EnumHand enumhand) throws InterruptedException {
         ItemStack itemstack = entityhuman.b(enumhand);
 
         if (itemstack.getItem() == Items.BOWL && this.getAge() >= 0 && !entityhuman.abilities.canInstantlyBuild) {
@@ -28,9 +29,9 @@ public class EntityMushroomCow extends EntityCow {
 
             return true;
         } else if (itemstack.getItem() == Items.SHEARS && this.getAge() >= 0) {
             // CraftBukkit start
             PlayerShearEntityEvent event = new PlayerShearEntityEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), this.getBukkitEntity());
+            event.wait(250);
             this.world.getServer().getPluginManager().callEvent(event);
 
             if (event.isCancelled()) {
@@ -39,7 +40,6 @@ public class EntityMushroomCow extends EntityCow {
             // CraftBukkit end
             this.die();
             // this.world.addParticle(EnumParticle.EXPLOSION_LARGE, this.locX, this.locY + (double) (this.length / 2.0F), this.locZ, 0.0D, 0.0D, 0.0D, new int[0]); // Reaper - Clientside particles
             if (true) { // Reaper - Remove isClientSide check
                 EntityCow entitycow = new EntityCow(this.world);
 
