From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Sun, 31 Jul 2022 11:41:39 +0500
Subject: [PATCH] fix: prevent vehicle move while awaiting teleport confirm


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 7bc594293b3f2e15fc937e30ebad48ef54d63ee1..ed7352cb23bcbfefb4b25627d8a8f0145e17079c 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -302,6 +302,16 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
             Entity entity = this.player.getVehicle();
 
             if (entity != this.player && entity.bE() == this.player && entity == this.r) {
+                // Reaper start - Prevent vehicle move while awaiting teleport confirm
+                if (this.teleportPos != null) {
+                    if (this.e - this.A > 20) {
+                        this.A = this.e;
+                        this.a(this.teleportPos.x, this.teleportPos.y, this.teleportPos.z, this.player.yaw, this.player.pitch);
+                    }
+                    this.allowedPlayerTicks = 20; // CraftBukkit
+                    return;
+                }
+                // Reaper end
                 WorldServer worldserver = this.player.x();
                 double d0 = entity.locX;
                 double d1 = entity.locY;
