From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Fri, 15 Jul 2022 00:10:35 +0500
Subject: [PATCH] perf-check-channel-before-reading


diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index 4467c9edce2425a91c7f4428b879e3645423b2c4..a7fb46f04dad4503951300e8e879bfee69723376 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -530,7 +530,9 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
     }
 
     protected void channelRead0(ChannelHandlerContext channelhandlercontext, Packet object) throws Exception { // CraftBukkit - fix decompile error
-        this.a(channelhandlercontext, (Packet) object);
+        if (isConnected()) { // Reaper - check if channel is opened before reading packet
+            this.a(channelhandlercontext, object);
+        }
     }
 
     static class QueuedPacket {
