From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Thu, 14 Jul 2022 23:43:44 +0500
Subject: [PATCH] perf-Use-valid-item-for-enchantment-checks-on-block


diff --git a/src/main/java/net/minecraft/server/EntityMushroomCow.java b/src/main/java/net/minecraft/server/EntityMushroomCow.java
index 1454dfa31445f381cf7bc0794e82c774288cb78b..5edbcec3bb7a8526943a10d260fbbab00385e722 100644
--- a/src/main/java/net/minecraft/server/EntityMushroomCow.java
+++ b/src/main/java/net/minecraft/server/EntityMushroomCow.java
@@ -38,6 +38,7 @@ public class EntityMushroomCow extends EntityCow {
             // CraftBukkit end
             this.die();
             // this.world.addParticle(EnumParticle.EXPLOSION_LARGE, this.locX, this.locY + (double) (this.length / 2.0F), this.locZ, 0.0D, 0.0D, 0.0D, new int[0]); // Reaper - Clientside particles
+            if (this.dead) return true; // Reaper - Fi-x cow dupe
             if (true) { // Reaper - Remove isClientSide check
                 EntityCow entitycow = new EntityCow(this.world);
 
diff --git a/src/main/java/net/minecraft/server/PlayerInteractManager.java b/src/main/java/net/minecraft/server/PlayerInteractManager.java
index 61c4a1e06bc3e8deee52ec1994d078c1aa5c1488..bd33f6f4bcca64e4b55787926ad2f5eb4a0879d3 100644
--- a/src/main/java/net/minecraft/server/PlayerInteractManager.java
+++ b/src/main/java/net/minecraft/server/PlayerInteractManager.java
@@ -399,9 +399,9 @@ public class PlayerInteractManager {
                     this.player.playerConnection.sendPacket(new PacketPlayOutBlockChange(this.world, blockposition));
                 } else {
                     ItemStack itemstack1 = this.player.getItemInMainHand();
-                    ItemStack itemstack2 = itemstack1.isEmpty() ? ItemStack.a : itemstack1.cloneItemStack();
+                    // ItemStack itemstack2 = itemstack1.isEmpty() ? ItemStack.a : itemstack1.cloneItemStack(); //Reaper - Move up
                     boolean flag1 = this.player.hasBlock(iblockdata);
-
+                    ItemStack itemstack2 = flag && flag1 && event.isDropItems() && !itemstack1.isEmpty() ? itemstack1.cloneItemStack() : ItemStack.a; // Reaper - clone before use
                     if (!itemstack1.isEmpty()) {
                         itemstack1.a(this.world, iblockdata, blockposition, this.player);
                     }
