From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Softik Lord <dimap9986@gmail.com>
Date: Fri, 15 Jul 2022 02:41:04 +0500
Subject: [PATCH] perf-disable-Paper-timings-by-default


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index 3d6e09907b9d416fc011c5d434903baede20ab49..cfeef132d51bccf1266d3a468fbec445ad5b649e 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -189,6 +189,15 @@ public class PaperConfig {
         int timingHistoryInterval = getInt("timings.history-interval", 300);
         int timingHistoryLength = getInt("timings.history-length", 3600);
 
+        // Reaper start
+        boolean reallyEnabled = getBoolean("timings.really-enabled", false);
+        if (timings && !reallyEnabled) {
+            Bukkit.getLogger().log(Level.WARNING, "[Reaper] To improve performance, timings have been disabled by default");
+            Bukkit.getLogger().log(Level.WARNING, "[Reaper] You can still use timings by using /timings on, but they will not start on server startup unless you set timings.really-enabled to true in paper.yml");
+            Bukkit.getLogger().log(Level.WARNING, "[Reaper] If you would like to disable this message, either set timings.really-enabled to true or timings.enabled to false.");
+        }
+        timings = reallyEnabled;
+        // Reaper end
 
         Timings.setVerboseTimingsEnabled(verboseTimings);
         Timings.setTimingsEnabled(timings);
