From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Thu, 27 Apr 2023 03:41:00 +0800
Subject: [PATCH] perf: cache random ticks


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index b1b24f5074e866a4615f5d900165be7965ae0a63..077e94ef41f7c023533d6ee6ce57a7cbce75e97f 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -122,6 +122,25 @@ public class Chunk {
     }
     // CraftBukkit end
 
+    // Reaper start - Cache random ticks
+    private int lightningTick;
+    boolean shouldDoLightning(Random random) {
+        if (this.lightningTick-- <= 0) {
+            this.lightningTick = random.nextInt(100000);
+            return true;
+        }
+        return false;
+    }
+    private int iceAndSnowTick;
+    boolean shouldDoIceAndSnow(Random random) {
+        if (this.iceAndSnowTick-- <= 0) {
+            this.iceAndSnowTick = random.nextInt(16);
+            return true;
+        }
+        return false;
+    }
+    // Reaper end
+
     public Chunk(World world, int i, int j) {
         this.sections = new ChunkSection[16];
         this.g = new byte[256];
@@ -142,6 +161,10 @@ public class Chunk {
 
         Arrays.fill(this.h, -999);
         Arrays.fill(this.g, (byte) -1);
+        // Reaper start - Cache random ticks
+        shouldDoLightning(world.random);
+        shouldDoIceAndSnow(world.random);
+        // Reaper end
         // CraftBukkit start
         this.bukkitChunk = new org.bukkit.craftbukkit.CraftChunk(this);
         this.chunkKey = ChunkCoordIntPair.a(this.locX, this.locZ);
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 0da618dec4afc13cf313d5b4193ca0db930040df..d0022fa497688aa10071d8a81c5fb672824490c9 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -509,7 +509,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
                 BlockPosition blockposition;
 
                 // Paper - Disable thunder
-                if (!this.paperConfig.disableThunder && flag && flag1 && this.random.nextInt(100000) == 0) {
+                if (!this.paperConfig.disableThunder && flag && flag1 && chunk.shouldDoLightning(this.random)) { // Reaper - Cache random ticks
                     this.l = this.l * 3 + 1013904223;
                     l = this.l >> 2;
                     blockposition = this.a(new BlockPosition(j + (l & 15), 0, k + (l >> 8 & 15)));
@@ -531,7 +531,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
                 }
 
                 // this.methodProfiler.c("iceandsnow"); // Reaper - Remove MethodProfiler
-                if (!this.paperConfig.disableIceAndSnow && this.random.nextInt(16) == 0) { // Paper - Disable ice and snow
+                if (!this.paperConfig.disableIceAndSnow && chunk.shouldDoIceAndSnow(this.random)) { // Paper - Disable ice and snow // Reaper - Cache random ticks
                     this.l = this.l * 3 + 1013904223;
                     l = this.l >> 2;
                     blockposition = this.p(new BlockPosition(j + (l & 15), 0, k + (l >> 8 & 15)));
