From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Tue, 18 Jul 2023 23:04:40 +0800
Subject: [PATCH] fix: packet click crash


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 329eb0dec15845e8f8f2df2afadd5df473929631..7aeb27be54a53ef63d12e05351b8b28ef4467d9d 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1889,6 +1889,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                         } else if (packetplayinwindowclick.c() == 1) {
                             click = ClickType.RIGHT;
                         }
+                        if (click == ClickType.UNKNOWN) return; // Reaper - Fix click crash
                         if (packetplayinwindowclick.c() == 0 || packetplayinwindowclick.c() == 1) {
                             action = InventoryAction.NOTHING; // Don't want to repeat ourselves
                             if (packetplayinwindowclick.b() == -999) {
