From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Tue, 18 Jul 2023 23:04:40 +0800
Subject: [PATCH] fix: packet click crash


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 329eb0dec15845e8f8f2df2afadd5df473929631..d4f6677dc68a67a7d7b580c8298c2332dff91ff4 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1872,6 +1872,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                 if (packetplayinwindowclick.b() < -1 && packetplayinwindowclick.b() != -999) {
                     return;
                 }
+                if (this.player.activeContainer.slots.size() <= packetplayinwindowclick.b()) return; // Reaper - Fix click crash
 
                 InventoryView inventory = this.player.activeContainer.getBukkitView();
                 SlotType type = CraftInventoryView.getSlotType(inventory, packetplayinwindowclick.b());
